### PR TITLE
[3/n] DP Enhancement: Padding tokens to max length when workload is balanced

### LINKED ISSUE #5527: Cuda graph supported bs in DP attention
I have been testing and recording the output throughput of SGLang on 2*8 H100 GPUs, and I've observed a significant regression in output throughput for long outputs in the `enable-dp-attention` scenarios following this [PR](https://github.com/sgl-project/sglang/pull/4390). Through debugging and profiling with Nsight Systems, I confirmed that the performance degradation is caused by the CUDA graph not being properly launched.

See the [code](https://github.com/sgl-project/sglang/blob/00391f58a3c2b13b00dc0a0a983a5c7fab399883/python/sglang/srt/model_executor/cuda_graph_runner.py#L280)
```
if self.enable_dp_attention:
    total_global_tokens = sum(forward_batch.global_num_tokens_cpu)

    is_bs_supported = forward_batch.can_run_dp_cuda_graph and (
        total_global_tokens in self.graphs
        if self.disable_padding
        else total_global_tokens <= self.max_bs
    )
```
After `enable-dp-attention`, `total_global_tokens` equals the sum of tokens across all DP ranks. For example, during the decode phase with DP = TP = 16 and a per-rank batch size of 32, the `total_global_tokens` would be 32 * 16 = 512. However, the maximum batch size allowed for CUDA graph capture defaults to 160. As a result, the `can_run` function during the decode phase returns `False`, and CUDA graph execution is consequently skipped.

In fact, according to the design logic of the [PR](https://github.com/sgl-project/sglang/pull/4390), I don't consider this a bug—it consistently uses the `total_global_tokens` across all ranks as the batch size to be captured by the CUDA graph. A straightforward solution would be to set a sufficiently large `cuda-graph-max-bs` when launching the server, though this might consume a significant amount of additional memory.

I believe using the `num_tokens` per DP rank as the CUDA graph batch size might be a more reasonable approach, similar to the code prior to this [PR](https://github.com/sgl-project/sglang/pull/4390). It would only require reserving adequate space for the `gathered_buffer`.

Below are my test output throughput before and after this [PR](https://github.com/sgl-project/sglang/pull/4390).
2*8 H100, input_len=output_len=1000, DP=TP=16
| Concurrency| before PR | after PR | after PR-fix |
|:---:|:---:|:---:|:---:|
| 1024  | 5115.13  | 2581.07  | 5469.53  |
| 512  | 3897.78  | 1527.95  | 4509.96  |

### LINKED ISSUE #7951: [Bug] Tensor shape is wrong when cudagraph+enable_dp_attention
### Checklist

- [ ] 1. I have searched related issues but cannot get the expected help.
- [ ] 2. The bug has not been fixed in the latest version.
- [ ] 3. Please note that if the bug-related issue you submitted lacks corresponding environment info and a minimal reproducible demo, it will be challenging for us to reproduce and resolve the issue, reducing the likelihood of receiving feedback.
- [ ] 4. If the issue you raised is not a bug but a question, please raise a discussion at https://github.com/sgl-project/sglang/discussions/new/choose Otherwise, it will be closed.
- [ ] 5. Please use English, otherwise it will be closed.

### Describe the bug

I tried to run DSR1 fp4 model on 8xB200, but found that some issue when I opened cudagraph and attndp, the input tensor dimension for each MoE layer is padded to global bs. For example, I take global bs 4096 and attention dp 8, which each rank should have 512 reqs for decode and the input tensor M dimension should be 512 for local rank. 
But I tried to do some profiling, I found that when cudagraph is on, each rank has input M dim 4096, not 512. When cudagraph is off, each rank has input M dim 512 which looks good.
Is this known or a bug?
Without cudagraph

Image

With cudagraph

Image

### Reproduction

**Server:**
python3 -m sglang.launch_server \
--model-path nvidia/DeepSeek-R1-0528-FP4 \
--trust-remote-code \
--quantization modelopt_fp4 \
--dp-size 8 --enable-dp-attention --enable-dp-lm-head\
--tp-size 8 \
--attention-backend cutlass_mla \
--enable-ep-moe \
--enable-flashinfer-moe \
--cuda-graph-bs 1 2 4 8 16 32 64 128 256 512 1024 2048 4096 \
--chunked-prefill-size 16384 \
--mem-fraction-static 0.85 \
--max-running-requests 4096 \
--stream-interval 5 
**Client:**
benchmark_serving.py with isl/osl 1024/1024, concurrency 4096.

### Environment

latest main.

### PR DESCRIPTION
**Pull request by ch-wan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #8279
* __->__ #8278
* #8277
* #8276

## Background

Integrating CUDA Graph with DP attention is tricky because different DP ranks have different workload. Simply gathering tensors from all DP ranks would incur dynamic shape. There are two design choices to make DP attention compatible with CUDA Graph:

- **Padding tokens to `max(len)` and use *all-gather* to gather tensors.** This design pads the input tensor of FFN to `max_len * dp_size` and incurs redundant computation (see examples in [#4390](https://github.com/sgl-project/sglang/pull/4390) and response [here](https://github.com/sgl-project/sglang/pull/6092#issuecomment-2860473971)).
- **Padding tokens to `sum(len)` and use *all-reduce* to gather tensors.** This design padded tokens to `sum(len)` which increases computation in the attention layers ([#6092](https://github.com/sgl-project/sglang/pull/6092) [#7951](https://github.com/sgl-project/sglang/issues/7951)). In addition, the cuda graph batch size is `sum(len)` rather than the actual running batch size on each DP worker. The users need to increase cuda graph batch size when DP size in creased, which causes a lot of confusion or misuse ([#5527](https://github.com/sgl-project/sglang/issues/5527) [#5557](https://github.com/sgl-project/sglang/issues/5557) [#6092](https://github.com/sgl-project/sglang/issues/6092)).

This PR offers two gather modes. When CUDA graph is disabled, we the select the option that minimizes communication. When CUDA graph is in use, we use the first option.

## Future Work

Under `max_len` padding mode, the input hidden_state of FFN is padded to `max_len * dp_size`, which may incur redundant computation. In the future, we can use masked (gouped-)GeMM kernels (e.g., DeepGEMM) to optimize it.

### CODE REVIEW

File: python/sglang/srt/layers/dp_attention.py [hunk 1]
```diff
@@ -30,6 +31,33 @@
...
+        return self == DPGatherMode.ALL_GATHER
+
+    def is_all_reduce(self):
+        return self == DPGatherMode.ALL_REDUCE
+
+    @classmethod
+    def get_dp_gather_mode(cls, global_num_tokens: List[int]) -> DPGatherMode:
+        max_len = max(global_num_tokens)
+        sum_len = sum(global_num_tokens)
+        if sum_len * 2 > max_len * get_attention_dp_size():
```
Comment (ispobock) [inline python/sglang/srt/layers/dp_attention.py:52]: Just curious, how did we get this formula?

File: python/sglang/srt/speculative/eagle_draft_extend_cuda_graph_runner.py [hunk 2]
```diff
@@ -306,12 +308,13 @@ def replay(self, forward_batch: ForwardBatch):
         raw_bs = forward_batch.batch_size
         num_tokens = forward_batch.input_ids.shape[0]
         if self.require_mlp_tp_gather:
-            total_batch_size = (
-                sum(forward_batch.global_num_tokens_cpu) // self.num_tokens_per_bs
+            max_num_tokens = max(forward_batch.global_num_tokens_cpu)
```
Comment (ispobock) [inline python/sglang/srt/speculative/eagle_draft_extend_cuda_graph_runner.py:311]: nit: use `forward_batch.global_num_tokens_cpu.max().item()` may be more efficient?

File: python/sglang/srt/model_executor/forward_batch_info.py [hunk 3]
```diff
@@ -585,34 +592,158 @@ def prepare_chunked_kv_indices(self, device: torch.device):
             )
             self.prefix_chunk_kv_indices.append(chunk_kv_indices)
 
-    def _pad_tensor_to_size(self, tensor: torch.Tensor, size: int):
-        return torch.cat(
-            [tensor, tensor.new_zeros(size - tensor.shape[0], *tensor.shape[1:])], dim=0
-        )
+    def _pad_tensor_to_size(self, tensor: torch.Tensor, size: int, *, value: int = 0):
+        if value == 0:
+            return torch.cat(
```
Comment (ispobock) [inline python/sglang/srt/model_executor/forward_batch_info.py:597]: nit: why don't use torch pad function?

**Review by ispobock — APPROVED** [review]
LGTM

File: python/sglang/srt/layers/dp_attention.py (hunk 1 above)
Comment (ch-wan) [reply python/sglang/srt/layers/dp_attention.py:52]: The communication volume of each worker for AllGather mode and AllReduce mode are `max_len * dp_size` and `sum_len * 2`, respectively.

File: python/sglang/srt/model_executor/forward_batch_info.py (hunk 3 above)
Comment (ch-wan) [reply python/sglang/srt/model_executor/forward_batch_info.py:597]: Are you asking this one (https://docs.pytorch.org/docs/stable/generated/torch.nn.functional.pad.html)? It is for image padding. Unfortunately, I'm not sure how to use them.

File: python/sglang/srt/speculative/eagle_draft_extend_cuda_graph_runner.py (hunk 2 above)
Comment (ch-wan) [reply python/sglang/srt/speculative/eagle_draft_extend_cuda_graph_runner.py:311]: Good catch. I'm going to revise this.

Comment (ch-wan) [reply python/sglang/srt/speculative/eagle_draft_extend_cuda_graph_runner.py:311]: lol, `global_num_tokens_cpu` is a list of int
